File: Hospital/src/hospital/Fila.java
package hospital;

class NoFila{
	
	protected String modalidade;
	protected int numero;
	protected NoFila proximo;
	
	public NoFila(int numero, String modalidade) {
		if(numero > 0) this.numero = numero;
		else System.exit(1);
		
		this.modalidade = modalidade;
	}
		
	public String getNo(){
		return this.numero + this.modalidade;
	}
}

public class Fila {
	
	private NoFila cabeca;
	private NoFila calda;
	private int senhaAtual;
	
	Fila(){
		this.cabeca = null;
		this.calda = null;
		this.senhaAtual = 0;
	}
	
	public String adicionar(String modalidade){
		senhaAtual++;
		
		NoFila novoNo = new NoFila(senhaAtual, modalidade);
		
		if(this.cabeca == null) {
			this.cabeca = novoNo;
			this.calda = novoNo;
		} else {
			this.calda.proximo = novoNo;
			this.calda = novoNo;
		}
		
		return novoNo.getNo();
	}
	
	public String chamar() {
		String resposta;
		
        if(this.cabeca == null){
        	resposta = "Fila Vazia";
        } else {
        	resposta = this.cabeca.getNo();
        	this.cabeca = this.cabeca.proximo;
        }
        return resposta;
	}

}

File: Hospital-Backend/Fila.js
class NoFila{
    constructor(valor, proximo){
        this.valor = valor;
        this.proximo = proximo;
    }
}

class Fila{
    constructor(){
        this.senhaAtual = 1;
        this.cabeça = null;
        this.calda = null;
    }

    add(modalidade){
        let novoNo = new NoFila(this.senhaAtual+modalidade, null);

        this.senhaAtual++;

        if(this.cabeça == null){
            this.cabeça = novoNo;
            this.calda = novoNo;
        } else {
            this.calda.proximo = novoNo;
            this.calda = novoNo;
        }

        return novoNo.valor;
    }

    remove(){
        let senhaRemovida = this.cabeça;
        if(this.cabeça == null){
            console.log("Fila Vazia");
            return "Fila Vazia";
        } else {
            this.cabeça = this.cabeça.proximo;
        }
        return senhaRemovida.valor;
    }

    ler(){
        let noAtual = this.cabeça;
        console.log("Lendo Fila")
        while(noAtual != null){
            console.log(noAtual.valor);
        }
    }

    esvaziar(){
        console.log("Evaziando");
        this.cabeça = null;
    }
}

module.exports = Fila;